### ISSUE TITLE
Abstract refinements + reflection + GADTs causes fixpoint crash

### ISSUE THREAD

**Issue by rosekunkel** [ucsd-progsys/liquidhaskell#1423]
The following file causes a crash:
```haskell
{-# LANGUAGE GADTs #-}
{-@ LIQUID "--reflection" @-}
{-@ LIQUID "--exact-data-con" @-}
module Gadt where

{-@
data Foo a <p :: Int -> Bool> where
   A :: Foo<{\_ -> True}> ()
@-}
data Foo a where
  A :: Foo ()

{-@ reflect getFoo @-}
getFoo :: Foo a -> a
getFoo A = ()
```
The output is
```text
LiquidHaskell Version 0.8.4.0, Git revision e557f2f4f992afcaf25dd70d1e73865334668138
Copyright 2013-18 Regents of the University of California. All Rights Reserved.


**** DONE:  A-Normalization ****************************************************
 

**** DONE:  Extracted Core using GHC *******************************************
 

**** DONE:  Transformed Core ***************************************************
 
liquid: Constraint.instantiatePvs
CallStack (from HasCallStack):
  error, called at src/Language/Fixpoint/Misc.hs:151:14 in liquid-fixpoint-0.8.0.1-5lBHQ0phVgG9Ya0ErK2yUa:Language.Fixpoint.Misc
```

**Comment by ranjitjhala** [comment]
Oh no. Regrettably this has touched the dark corners of the abstract
refinement implementation that only @nikivazou (if anyone) understands ...

**Comment by nikivazou** [comment]
Well, I have no idea how are GADTs represented in core and how they are different than ADTs. 
Will take a look next week.

**Comment by ranjitjhala** [comment]
Yes it’s quite likely that the GADTs you some different kind of
representation and that the code for abstract refinements need to to be
updated to account for that...

**Comment by nikivazou** [comment]
Fixed, but check my comment in #1439
